bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java

```diff
@@ -0,0 +1,492 @@
+        }
+
+        if (!endpoint.isPresent()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE, String.format(
+                    "Bridge '%s' not completely initialized", Optional.ofNullable(getBridge()).map(b -> b.getLabel())));
+            logger.debug("Bridge not initialized fully (no endpoint) -- aborting init for {}", this);
+            return;
+        }
+    }
+
+    /**
+     * Remove the endpoint if exists
+     */
+    private synchronized void unregisterEndpoint() {
+        if (endpoint.isPresent()) {
+            endpoint = Optional.empty();
+        }
+    }
+
+    /**
+     * Register poll task
+     * This is where we set up our regular poller
+     */
+    private synchronized void registerPollTask(ModelBlock mainBlock) {
+        if (pollTask.isPresent()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+            throw new IllegalStateException("pollTask should be unregistered before registering a new one!");
+        }
+        if (!config.isPresent() || !endpoint.isPresent()) {
+            throw new IllegalStateException("registerPollTask called without proper configuration");
+        }
+
+        logger.debug("Setting up regular polling");
+
+        BasicModbusReadRequestBlueprint request = new BasicModbusReadRequestBlueprint(getSlaveId(),
+                ModbusReadFunctionCode.READ_MULTIPLE_REGISTERS, mainBlock.address, mainBlock.length,
+                config.get().getMaxTries());
+
+        pollTask = Optional.of(new BasicPollTaskImpl(endpoint.get(), request, new ModbusReadCallback() {
+
+            @Override
+            public void onRegisters(@Nullable ModbusReadRequestBlueprint request,
+                    @Nullable ModbusRegisterArray registers) {
+                if (registers == null) {
+                    logger.info("Received empty register array on poll");
+                    return;
+                }
+
+                handlePolledData(registers);
+
+                if (getThing().getStatus() != ThingStatus.ONLINE) {
+                    updateStatus(ThingStatus.ONLINE);
+                }
+            }
+
+            @Override
+            public void onError(@Nullable ModbusReadRequestBlueprint request, @Nullable Exception error) {
+                handleError(error);
+            }
+
+            @Override
+            public void onBits(@Nullable ModbusReadRequestBlueprint request, @Nullable BitArray bits) {
+                // don't care, we don't expect this result
+            }
+        }));
+
+        managerRef.registerRegularPoll(pollTask.get(), config.get().getRefreshMillis(), 1000);
+    }
+
+    /**
+     * This method should handle incoming poll data, and update the channels
+     * with the values received
+     */
+    protected abstract void handlePolledData(ModbusRegisterArray registers);
+
+    @Override
+    public synchronized void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
```

Comment:
why is this synchronized?

Reply:
Good point. I was looking up these things from another bundle, but as I see it's not synchronized in the parent class either.
Changed to non-synchronized.